test/extensions/key_value/file_based/key_value_store_test.cc

```diff
@@ -103,6 +103,33 @@ TEST_F(KeyValueStoreTest, Iterate) {
   EXPECT_EQ(1, stop_early_counter);
 }
 
+#ifndef NDEBUG
+TEST_F(KeyValueStoreTest, ShouldCrashIfIterateCallbackAddsOrUpdatesStore) {
+  store_->addOrUpdate("foo", "bar");
+  store_->addOrUpdate("baz", "eep");
+  KeyValueStore::ConstIterateCb update_value_callback = [this](const std::string& key,
+                                                               const std::string&) {
+    EXPECT_TRUE(key == "foo" || key == "baz");
+    store_->addOrUpdate("foo", "updated-bar");
+    return KeyValueStore::Iterate::Continue;
+  };
+
+  EXPECT_DEATH(store_->iterate(update_value_callback),
+               "Expected iterate to not modify the underlying store.");
+
+  KeyValueStore::ConstIterateCb add_key_callback = [this](const std::string& key,
+                                                          const std::string&) {
+    EXPECT_TRUE(key == "foo" || key == "baz" || key == "new-key");
+    if (key == "foo") {
+      store_->addOrUpdate("new-key", "new-value");
+    }
+    return KeyValueStore::Iterate::Continue;
+  };
+  EXPECT_DEATH(store_->iterate(add_key_callback),
+               "Expected iterate to not modify the underlying store.");
+}
```

Comment:
That seems reasonable. Another option to consider would be adding a member like num_running_iterations_ which was incremented when iterate() was called, and decremented when it finished. Then each of the methods which modify the store could ASSERT(num_running_iterations_ == 0). (All of that wrapped in the NDEBUG guard). This would allow an erase test to be written and would have a clear assert failure if erase() were called during iteration. But I'm not sure if that's worth doing?